ln-gateway/src/cln.rs

```diff
@@ -40,3 +62,251 @@ pub struct HtlcAccepted {
+                "gw-withdraw",
+                "Withdraw from federation. Args: <address> <sats>",
+                withdraw_rpc,
+            )
+            .rpcmethod("gw-address", "Generate deposit address", address)
+            .hook("htlc_accepted", |plugin, value| async move {
+                // This callback needs to be `Sync`, so we use tokio::spawn
+                let handle = tokio::spawn(async move {
+                    htlc_accepted_hook(plugin, value).await.or_else(|e| {
+                        error!("htlc_accepted error {:?}", e);
+                        // cln_plugin doesn't handle errors very well ... tell it to proceed normally
+                        Ok(json!({ "result": "continue" }))
+                    })
+                });
+                handle.await?
+            })
+            .dynamic() // Allow reloading the plugin
+            .start()
+            .await?
+        {
+            Ok(plugin)
+        } else {
+            Err(anyhow!("Failed to build cln rpc!"))
+        }
+    }
+
+    async fn build_cln_rpc(
+        &self,
+        messenger: GatewayMessageSender,
+    ) -> Result<(Arc<dyn LnRpc>, SocketAddr, PathBuf), Error> {
+        let plugin = self.build_cln_plugin(messenger).await?;
+
+        let workdir = match plugin.option("fedimint-cfg") {
+            Some(options::Value::String(workdir)) => {
+                // FIXME: cln_plugin doesn't yet support optional parameters
+                if &workdir == "default-dont-use" {
+                    panic!("fedimint-cfg option missing")
+                } else {
+                    PathBuf::from(workdir)
+                }
+            }
+            _ => unreachable!(),
+        };
+        let host = match plugin.option("fedimint-host") {
+            Some(options::Value::String(host)) => host,
+            _ => unreachable!(),
+        };
+        let port = match plugin.option("fedimint-port") {
+            Some(options::Value::String(port)) => port,
+            _ => unreachable!(),
+        };
+        let bind_addr = format!("{}:{}", host, port)
+            .parse()
+            .expect("Invalid gateway bind address");
+
+        // If no config exists, try to generate one
+        let cfg_path = workdir.join("gateway.json");
+        let config = plugin.configuration();
+        let cln_rpc_socket = PathBuf::from(config.lightning_dir).join(config.rpc_file);
+        let mut ln_client = ClnRpc::new(cln_rpc_socket)
+            .await
+            .expect("connect to ln_socket");
+        if !Path::new(&cfg_path).is_file() {
+            generate_config(&workdir, &mut ln_client, &bind_addr).await;
+        }
+        Ok((Arc::new(Mutex::new(ln_client)), bind_addr, workdir))
+    }
+}
+
+#[async_trait]
+impl LnRpcFactory for ClnRpcFactory {
+    async fn create(
+        &self,
+        messenger: GatewayMessageSender,
+    ) -> Result<(Arc<dyn LnRpc>, SocketAddr, PathBuf), Error> {
+        match self.build_cln_rpc(messenger).await {
```

Comment:
Yo @dpc . help me make this snippet thread safe? ATM I'm not sure if I really need to pass `LnRpcFactory` across threads. In any case, I get a warning about this future call not being thread safe.
```
future cannot be sent between threads safely
within `impl futures::Future<Output = std::result::Result<(Arc<dyn LnRpc>, std::net::SocketAddr, PathBuf), anyhow::Error>>`, the trait `std::marker::Send` is not implemented for `*mut rand::rngs::adapter::ReseedingRng<rand_hc::hc128::Hc128Core, rand::rngs::EntropyRng>`
required for the cast from `impl futures::Future<Output = std::result::Result<(Arc<dyn LnRpc>, std::net::SocketAddr, PathBuf), anyhow::Error>>` to the object type `dyn futures::Future<Output = std::result::Result<(Arc<(dyn LnRpc + 'static)>, std::net::SocketAddr, PathBuf), anyhow::Error>> + std::marker::Send`rustc
cln.rs(178, 11): future is not `Send` as this value is used across an await
```

Reply:
Since tokio is a multithreaded async executor it needs to be able to move futures (`async` functions) between threads. For this they need to be thread-safe. `thread_rng` created `ThreadRng` which is not thread-safe: https://docs.rs/rand/0.6.5/rand/rngs/struct.ThreadRng.html#synthetic-implementations

We need to use something else. Probably `StdRng` but heaven forbids if `rand` made this actually easy ...

Reply:
```diff
> git diff HEAD
diff --git a/ln-gateway/src/cln.rs b/ln-gateway/src/cln.rs
index 1fc3308f5..c02e5ed77 100644
--- a/ln-gateway/src/cln.rs
+++ b/ln-gateway/src/cln.rs
@@ -5,7 +5,7 @@ use fedimint_api::Amount;
 use fedimint_server::config::load_from_file;
 use fedimint_server::modules::ln::contracts::Preimage;
 use mint_client::GatewayClientConfig;
-use rand::thread_rng;
+use rand::FromEntropy;
 use secp256k1::KeyPair;
 use serde::{Deserialize, Deserializer};
 use serde_json::json;
@@ -168,7 +168,7 @@ async fn generate_config(workdir: &Path, ln_client: &mut ClnRpc, bind_addr: &Soc
     let client_cfg_path = workdir.join("client.json");
     let client_cfg: fedimint_server::config::ClientConfig = load_from_file(&client_cfg_path);
 
-    let mut rng = thread_rng();
+    let mut rng = rand::rngs::StdRng::from_entropy();
     let ctx = secp256k1::Secp256k1::new();
     let kp_fed = KeyPair::new(&ctx, &mut rng);
```

This will do. `from_entropy` is just "hidden" on a trait, and had to google it to figure it out.

Reply:
Oh good pointer! Ps, I do realize that I can keep rng if I move the config gen stuff out of the direct path of building cln rpc. It's a bitgger change, but helps me formalize the two gateway in a much better way.

See : https://github.com/fedimint/fedimint/pull/709/commits/233769f65bd829ce14f270701fb13b17d66bd1e0